ddl/ddl_api.go

```diff
@@ -502,8 +502,12 @@ func isExplicitTimeStamp() bool {
 
 // processColumnFlags is used by columnDefToCol and processColumnOptions. It is intended to unify behaviors on `create/add` and `modify/change` statements. Check tidb#issue#19342.
 func processColumnFlags(col *table.Column) {
-	if col.FieldType.EvalType().IsStringKind() && col.Charset == charset.CharsetBin {
-		col.Flag |= mysql.BinaryFlag
+	if col.FieldType.EvalType().IsStringKind() {
```

Comment:
The flag is wrong after parsing, we need to correct it in the parser.

Reply:
I think there is nothing "wrong" about the binary flag. We need this flag to tell the difference between `(a varchar(255) binary)` and `(a varchar(255))`.

Reply:
> I think there is nothing "wrong" about the binary flag. We need this flag to tell the difference between `(a varchar(255) binary)` and `(a varchar(255))`.

`binary` in `varchar(255) binary` means `_bin`.  `bianry` flag means binary charset. They are not the same.

Reply:
If the binary flag only means binary charset. Why do we need this flag?

Reply:
@tangenta Yeah, good question. I check it again and I was wrong. Both _bin and binary have the binary flag.